go/genkit/session.go

```diff
@@ -0,0 +1,298 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package genkit
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"sync"
+
+	"github.com/firebase/genkit/go/ai"
+	"github.com/firebase/genkit/go/internal/base"
+	"github.com/google/uuid"
+	"github.com/invopop/jsonschema"
+)
+
+// Sessions are by default stored in memory only.
+// Implement the SessionStore interface to persist state.
+type SessionStore interface {
+	Get(sessionId string) (data SessionData, err error)
+	Save(sessionId string, data SessionData) error
+}
+
+type SessionData struct {
+	// Any state that should be stored
+	State map[string]any `json:"state,omitempty"`
+	// Schema for state variables
+	StateSchema *jsonschema.Schema `json:"stateschema,omitempty"`
+	// The messages for each thread
+	Threads map[string][]*ai.Message `json:"threads,omitempty"`
+}
+
+type Session struct {
+	// Genkit
+	Genkit *Genkit
+	// The session id
+	ID string
+	// The data for the session
+	SessionData SessionData
+	// Default state variable values
+	DefaultState map[string]any
+	// The store for the session, defaults to in-memory storage
+	Store SessionStore
+}
+
+// SessionOption configures params for the session
+type SessionOption func(s *Session) error
+
+// A session key
+type sessionKey string
+
+const (
+	currentSessionID    sessionKey = "currentSessionID"
+	currentSessionStore sessionKey = "currentSessionStore"
+)
+
+// NewSession creates a new session with the provided options.
+// If no store is provided, it defaults to in-memory storage.
+func NewSession(ctx context.Context, opts ...SessionOption) (session *Session, err error) {
+	s := &Session{}
+
+	for _, with := range opts {
+		err := with(s)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if s.Store == nil {
+		s.Store = &InMemSessionStore{}
+	}
+
+	if s.ID == "" {
+		s.ID = uuid.New().String()
+	}
+
+	if s.SessionData.Threads == nil {
+		s.SessionData.Threads = make(map[string][]*ai.Message)
+	}
+
+	// Only update state with defaults if not already set, eg. via WithSessionData
+	if s.SessionData.State == nil {
+		s.SessionData.State = s.DefaultState
+	}
+
+	// Initialize session in store
+	s.Store.Save(s.ID, s.SessionData)
+
+	return s, nil
+}
+
+// LoadSession loads sessiondata from store, and returns a session.
+func LoadSession(ctx context.Context, sessionId string, store SessionStore) (session *Session, err error) {
+	sessionData, err := store.Get(sessionId)
+	if err != nil {
+		return nil, err
+	}
+
+	session = &Session{
+		ID:          sessionId,
+		SessionData: sessionData,
+		Store:       store,
+	}
+
+	return session, nil
+}
+
+// UpdateState takes any data and sets it as state in the store.
+func (s *Session) UpdateState(state any) error {
```

Comment:
This looks like it modifies the state schema as well which we don't want to be doing. We should validate that the type matches the expected schema and update the values only.